File: packages/php/overlay/MgxClient.php
<?php

/* Hand-written overlay — not generated. */

declare(strict_types=1);

namespace MyGrainExchange\Mgx\Overlay;

use MyGrainExchange\Mgx\Api\BidsApi;
use MyGrainExchange\Mgx\Api\CashBidsApi;
use MyGrainExchange\Mgx\Api\InventoryApi;
use MyGrainExchange\Mgx\Api\MarketApi;
use MyGrainExchange\Mgx\Api\TeamsApi;
use MyGrainExchange\Mgx\Api\TradesApi;
use MyGrainExchange\Mgx\Api\WebhooksApi;
use MyGrainExchange\Mgx\ApiException;
use MyGrainExchange\Mgx\Configuration;

/**
 * The MGX Enterprise API client. Wraps the generated resource APIs with OAuth2
 * token handling, auto-pagination, idempotency keys, and typed errors.
 *
 * Resource namespaces are exposed via accessor methods that mirror the other
 * MGX SDKs: {@see inventory()}, {@see market()}, {@see bids()}, {@see trades()},
 * {@see teams()}, {@see cashBids()}, {@see webhooks()}.
 */
class MgxClient
{
    public const DEFAULT_BASE_URL = 'https://api.mygrainexchange.com/v1';

    private InventoryResource $inventory;
    private MarketResource $market;
    private BidsResource $bids;
    private TradesResource $trades;
    private TeamsResource $teams;
    private CashBidsResource $cashBids;
    private WebhooksResource $webhooks;

    /**
     * @param string|null       $clientId     OAuth2 client id (client-credentials or auth-code app).
     * @param string|null       $clientSecret OAuth2 client secret.
     * @param list<string>      $scopes       Scopes to request on a client-credentials grant.
     * @param string|null       $accessToken  A pre-obtained access token (e.g. from Login-with-MGX).
     * @param string|null       $refreshToken A refresh token to renew an auth-code access token.
     * @param string            $baseUrl      API base URL. Defaults to production.
     * @param string|null       $tokenUrl     Override the token endpoint (defaults to `<host>/oauth/token`).
     */
    public function __construct(
        ?string $clientId = null,
        ?string $clientSecret = null,
        array $scopes = [],
        ?string $accessToken = null,
        ?string $refreshToken = null,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?string $tokenUrl = null,
    ) {
        $tokens = new TokenManager(
            [
                'clientId' => $clientId,
                'clientSecret' => $clientSecret,
                'scopes' => $scopes,
                'accessToken' => $accessToken,
                'refreshToken' => $refreshToken,
                'tokenUrl' => $tokenUrl,
            ],
            $baseUrl,
        );

        // GOTCHA: the generated Api classes build the header as
        //   'Authorization: Bearer ' . $config->getAccessToken()
        // i.e. they add the "Bearer " prefix themselves. So the Configuration must
        // return the RAW token (not "Bearer <token>"), or the header would become
        // "Bearer Bearer <token>". TokenAwareConfiguration resolves the token lazily
        // on each getAccessToken() call so refreshes are picked up per request.
        $config = new TokenAwareConfiguration($tokens);
        $config->setHost(rtrim($baseUrl, '/'));

        $this->inventory = new InventoryResource(new InventoryApi(null, $config), new BidsApi(null, $config));
        $this->market = new MarketResource(new MarketApi(null, $config));
        $this->bids = new BidsResource(new BidsApi(null, $config));
        $this->trades = new TradesResource(new TradesApi(null, $config));
        $this->teams = new TeamsResource(new TeamsApi(null, $config));
        $this->cashBids = new CashBidsResource(new CashBidsApi(null, $config));
        $this->webhooks = new WebhooksResource(new WebhooksApi(null, $config));
    }

    /**
     * Convenience named-args factory mirroring the TS `new MgxClient({ ... })`
     * options object. Accepts the same keys as the constructor parameters.
     *
     * @param array{
     *   clientId?: string|null,
     *   clientSecret?: string|null,
     *   scopes?: list<string>,
     *   accessToken?: string|null,
     *   refreshToken?: string|null,
     *   baseUrl?: string,
     *   tokenUrl?: string|null
     * } $options
     */
    public static function create(array $options): self
    {
        return new self(
            clientId: $options['clientId'] ?? null,
            clientSecret: $options['clientSecret'] ?? null,
            scopes: $options['scopes'] ?? [],
            accessToken: $options['accessToken'] ?? null,
            refreshToken: $options['refreshToken'] ?? null,
            baseUrl: $options['baseUrl'] ?? self::DEFAULT_BASE_URL,
            tokenUrl: $options['tokenUrl'] ?? null,
        );
    }

    /** Browse and filter anonymized inventory. */
    public function inventory(): InventoryResource
    {
        return $this->inventory;
    }

    /** Market commodities, prices, and history. */
    public function market(): MarketResource
    {
        return $this->market;
    }

    /** Read the authenticated team's own bids and act on counter-offers. */
    public function bids(): BidsResource
    {
        return $this->bids;
    }

    /** Read the authenticated team's own trades. */
    public function trades(): TradesResource
    {
        return $this->trades;
    }

    /** Read the teams the authenticated user belongs to. */
    public function teams(): TeamsResource
    {
        return $this->teams;
    }

    /** Manage the elevator's own cash bids and the offers received on them. */
    public function cashBids(): CashBidsResource
    {
        return $this->cashBids;
    }

    /** Subscribe to event notifications and inspect deliveries. */
    public function webhooks(): WebhooksResource
    {
        return $this->webhooks;
    }
}

/**
 * A {@see Configuration} whose access token is resolved lazily from a
 * {@see TokenManager} on every read, so each request uses a fresh token.
 *
 * @internal
 */
final class TokenAwareConfiguration extends Configuration
{
    public function __construct(private readonly TokenManager $tokens)
    {
        parent::__construct();
    }

    public function getAccessToken()
    {
        // Returns the RAW token; the generated client adds the "Bearer " prefix.
        return $this->tokens->accessToken();
    }
}

/**
 * Shared helpers for resource wrappers.
 *
 * @internal
 */
trait ResourceHelpers
{
    /**
     * Run a generated API call, translating its ApiException into MgxApiError.
     *
     * @template R
     * @param callable(): R $call
     * @return R
     */
    private function call(callable $call)
    {
        try {
            return $call();
        } catch (ApiException $e) {
            throw MgxApiError::fromApiException($e);
        }
    }

    /** Generate a RFC 4122 version-4 UUID for idempotency keys. */
    private static function uuidV4(): string
    {
        $data = random_bytes(16);
        $data[6] = chr((ord($data[6]) & 0x0f) | 0x40);
        $data[8] = chr((ord($data[8]) & 0x3f) | 0x80);

        return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
    }
}

/** @internal */
final class InventoryResource
{
    use ResourceHelpers;

    public function __construct(private readonly InventoryApi $api, private readonly BidsApi $bidsApi)
    {
    }

    /**
     * Auto-paginating list of anonymized inventory lots.
     *
     * @param array<string, mixed> $filters commodity, grade, minQuantity, maxQuantity,
     *                                       province, near, cropYear, isOrganic, minProtein,
     *                                       sort, limit
     * @return Paginator<\MyGrainExchange\Mgx\Model\Inventory>
     */
    public function list(array $filters = []): Paginator
    {
        return new Paginator(fn (int $offset) => $this->call(fn () => $this->api->inventoryList(
            $filters['commodity'] ?? null,
            $filters['grade'] ?? null,
            $filters['minQuantity'] ?? null,
            $filters['maxQuantity'] ?? null,
            $filters['province'] ?? null,
            $filters['near'] ?? null,
            $filters['cropYear'] ?? null,
            $filters['isOrganic'] ?? null,
            $filters['minProtein'] ?? null,
            $filters['sort'] ?? null,
            $filters['limit'] ?? 20,
            $offset,
        )));
    }

    public function get(string $id): ?\MyGrainExchange\Mgx\Model\Inventory
    {
        return $this->call(fn () => $this->api->inventoryGet($id))->getData();
    }

    public function filters(): \MyGrainExchange\Mgx\Model\InventoryFilters
    {
        return $this->call(fn () => $this->api->inventoryFilters());
    }

    /**
     * Place a bid on an inventory lot. Auto-generates an Idempotency-Key when one
     * is not supplied.
     *
     * @param array<string, mixed>|\MyGrainExchange\Mgx\Model\PlaceBid $body
     */
    public function placeBid(string $id, $body, ?string $idempotencyKey = null): ?\MyGrainExchange\Mgx\Model\Bid
    {
        $key = $idempotencyKey ?? self::uuidV4();

        return $this->call(fn () => $this->bidsApi->bidPlace($id, $body, $key))->getData();
    }
}

/** @internal */
final class MarketResource
{
    use ResourceHelpers;

    public function __construct(private readonly MarketApi $api)
    {
    }

    /** @return list<\MyGrainExchange\Mgx\Model\MarketCommodity> */
    public function commodities(): array
    {
        return $this->call(fn () => $this->api->marketCommodities())->getItems() ?? [];
    }

    /**
     * @param array<string, mixed> $params commodity, date
     * @return list<\MyGrainExchange\Mgx\Model\MarketPrice>
     */
    public function prices(array $params = []): array
    {
        return $this->call(fn () => $this->api->marketPrices(
            $params['commodity'] ?? null,
            $params['date'] ?? null,
        ))->getItems() ?? [];
    }

    /**
     * @param array<string, mixed> $params from, to, interval
     */
    public function history(string $commodity, array $params = []): \MyGrainExchange\Mgx\Model\PriceHistory
    {
        return $this->call(fn () => $this->api->marketHistory(
            $commodity,
            $params['from'] ?? null,
            $params['to'] ?? null,
            $params['interval'] ?? 'day',
        ));
    }
}

/** @internal */
final class BidsResource
{
    use ResourceHelpers;

    public function __construct(private readonly BidsApi $api)
    {
    }

    /**
     * @param array<string, mixed> $params status, limit
     * @return Paginator<\MyGrainExchange\Mgx\Model\Bid>
     */
    public function list(array $params = []): Paginator
    {
        return new Paginator(fn (int $offset) => $this->call(fn () => $this->api->bidList(
            $params['status'] ?? null,
            $params['limit'] ?? 20,
            $offset,
        )));
    }

    public function get(string $id): ?\MyGrainExchange\Mgx\Model\Bid
    {
        return $this->call(fn () => $this->api->bidGet($id))->getData();
    }

    public function accept(string $id): ?\MyGrainExchange\Mgx\Model\Trade
    {
        return $this->call(fn () => $this->api->bidAccept($id))->getData();
    }

    public function reject(string $id): ?\MyGrainExchange\Mgx\Model\Bid
    {
        return $this->call(fn () => $this->api->bidReject($id))->getData();
    }

    /**
     * @param array<string, mixed>|\MyGrainExchange\Mgx\Model\PlaceBid $body
     */
    public function counter(string $id, $body): ?\MyGrainExchange\Mgx\Model\Bid
    {
        return $this->call(fn () => $this->api->bidCounter($id, $body))->getData();
    }
}

/** @internal */
final class TradesResource
{
    use ResourceHelpers;

    public function __construct(private readonly TradesApi $api)
    {
    }

    /**
     * @param array<string, mixed> $params status, commodity, from, to, limit
     * @return Paginator<\MyGrainExchange\Mgx\Model\Trade>
     */
    public function list(array $params = []): Paginator
    {
        return new Paginator(fn (int $offset) => $this->call(fn () => $this->api->tradeList(
            $params['status'] ?? null,
            $params['commodity'] ?? null,
            $params['from'] ?? null,
            $params['to'] ?? null,
            $params['limit'] ?? 20,
            $offset,
        )));
    }

    public function get(string $id): ?\MyGrainExchange\Mgx\Model\Trade
    {
        return $this->call(fn () => $this->api->tradeGet($id))->getData();
    }
}

/** @internal */
final class TeamsResource
{
    use ResourceHelpers;

    public function __construct(private readonly TeamsApi $api)
    {
    }

    /** @return list<\MyGrainExchange\Mgx\Model\Team> */
    public function list(): array
    {
        return $this->call(fn () => $this->api->teamList())->getItems() ?? [];
    }

    public function get(string $id): ?\MyGrainExchange\Mgx\Model\Team
    {
        return $this->call(fn () => $this->api->teamGet($id))->getData();
    }
}

/** @internal */
final class CashBidsResource
{
    use ResourceHelpers;

    public function __construct(private readonly CashBidsApi $api)
    {
    }

    /**
     * @param array<string, mixed> $params isActive, commodity, limit
     * @return Paginator<\MyGrainExchange\Mgx\Model\CashBid>
     */
    public function list(array $params = []): Paginator
    {
        return new Paginator(fn (int $offset) => $this->call(fn () => $this->api->cashBidList(
            $params['isActive'] ?? null,
            $params['commodity'] ?? null,
            $params['limit'] ?? 20,
            $offset,
        )));
    }

    /**
     * @param array<string, mixed>|\MyGrainExchange\Mgx\Model\StoreCashBid $body
     */
    public function create($body): ?\MyGrainExchange\Mgx\Model\CashBid
    {
        return $this->call(fn () => $this->api->cashBidCreate($body))->getData();
    }

    /**
     * @param array<string, mixed>|\MyGrainExchange\Mgx\Model\UpdateCashBid $body
     */
    public function update(string $id, $body): ?\MyGrainExchange\Mgx\Model\CashBid
    {
        return $this->call(fn () => $this->api->cashBidUpdate($id, $body))->getData();
    }

    /**
     * @param array<string, mixed> $params status, limit
     * @return Paginator<\MyGrainExchange\Mgx\Model\CashBidOffer>
     */
    public function offers(string $cashBidId, array $params = []): Paginator
    {
        return new Paginator(fn (int $offset) => $this->call(fn () => $this->api->cashBidOffers(
            $cashBidId,
            $params['status'] ?? null,
            $params['limit'] ?? 20,
            $offset,
        )));
    }

    public function acceptOffer(string $offerId): ?\MyGrainExchange\Mgx\Model\Trade
    {
        return $this->call(fn () => $this->api->cashBidOfferAccept($offerId))->getData();
    }

    public function rejectOffer(string $offerId): ?\MyGrainExchange\Mgx\Model\CashBidOffer
    {
        return $this->call(fn () => $this->api->cashBidOfferReject($offerId))->getData();
    }
}

/** @internal */
final class WebhooksResource
{
    use ResourceHelpers;

    public function __construct(private readonly WebhooksApi $api)
    {
    }

    /** @return list<\MyGrainExchange\Mgx\Model\WebhookResource> */
    public function list(): array
    {
        return $this->call(fn () => $this->api->webhookList())->getItems() ?? [];
    }

    /**
     * @param array<string, mixed>|\MyGrainExchange\Mgx\Model\Webhook $body
     */
    public function create($body): ?\MyGrainExchange\Mgx\Model\WebhookCreated
    {
        return $this->call(fn () => $this->api->webhookCreate($body))->getData();
    }

    public function delete(string $id): void
    {
        $this->call(fn () => $this->api->webhookDelete($id));
    }

    /**
     * @param array<string, mixed> $params limit
     * @return Paginator<\MyGrainExchange\Mgx\Model\WebhookDelivery>
     */
    public function deliveries(string $webhookId, array $params = []): Paginator
    {
        return new Paginator(fn (int $offset) => $this->call(fn () => $this->api->webhookDeliveries(
            $webhookId,
            $params['limit'] ?? 20,
            $offset,
        )));
    }

    /**
     * Verify an inbound webhook's signature and return the typed event.
     *
     * Pass the EXACT raw request body (not a re-serialized object) and the value
     * of the `MGX-Signature` header. Throws {@see MgxSignatureError} on a bad
     * signature or a stale timestamp.
     *
     * @param int|null $now Override "now" (unix seconds) — for testing.
     *
     * @throws MgxSignatureError
     */
    public function verify(
        string $rawBody,
        string $signatureHeader,
        string $secret,
        int $toleranceSeconds = DEFAULT_WEBHOOK_TOLERANCE_SECONDS,
        ?int $now = null,
    ): \MyGrainExchange\Mgx\Model\WebhookEvent {
        return verifyWebhook($rawBody, $signatureHeader, $secret, $toleranceSeconds, $now);
    }
}

File: packages/java/src/main/java/com/mygrainexchange/mgx/overlay/MgxClient.java
/* Hand-written overlay — not generated. */
package com.mygrainexchange.mgx.overlay;

import com.mygrainexchange.mgx.ApiClient;
import com.mygrainexchange.mgx.ApiException;
import com.mygrainexchange.mgx.api.BidsApi;
import com.mygrainexchange.mgx.api.CashBidsApi;
import com.mygrainexchange.mgx.api.InventoryApi;
import com.mygrainexchange.mgx.api.MarketApi;
import com.mygrainexchange.mgx.api.TeamsApi;
import com.mygrainexchange.mgx.api.TradesApi;
import com.mygrainexchange.mgx.api.WebhooksApi;
import com.mygrainexchange.mgx.model.Bid;
import com.mygrainexchange.mgx.model.BidPage;
import com.mygrainexchange.mgx.model.CashBid;
import com.mygrainexchange.mgx.model.CashBidOffer;
import com.mygrainexchange.mgx.model.CashBidOfferPage;
import com.mygrainexchange.mgx.model.CashBidPage;
import com.mygrainexchange.mgx.model.Inventory;
import com.mygrainexchange.mgx.model.InventoryPage;
import com.mygrainexchange.mgx.model.InventoryFilters;
import com.mygrainexchange.mgx.model.MarketCommodity;
import com.mygrainexchange.mgx.model.MarketPrice;
import com.mygrainexchange.mgx.model.PlaceBid;
import com.mygrainexchange.mgx.model.PriceHistory;
import com.mygrainexchange.mgx.model.StoreCashBid;
import com.mygrainexchange.mgx.model.Team;
import com.mygrainexchange.mgx.model.Trade;
import com.mygrainexchange.mgx.model.TradePage;
import com.mygrainexchange.mgx.model.UpdateCashBid;
import com.mygrainexchange.mgx.model.Webhook;
import com.mygrainexchange.mgx.model.WebhookCreated;
import com.mygrainexchange.mgx.model.WebhookDelivery;
import com.mygrainexchange.mgx.model.WebhookDeliveryPage;
import com.mygrainexchange.mgx.model.WebhookEvent;
import com.mygrainexchange.mgx.model.WebhookResource;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * The MGX Enterprise API client. Wraps the generated resource APIs with OAuth2
 * token handling, auto-pagination, idempotency keys, and typed errors.
 *
 * <pre>{@code
 * MgxClient mgx = MgxClient.builder()
 *     .clientId("...")
 *     .clientSecret("...")
 *     .scopes("inventory.read", "bids.write")
 *     .build();
 *
 * for (Inventory lot : mgx.inventory().list(new InventoryListFilters().commodity("wheat"))) {
 *     System.out.println(lot.getId());
 * }
 * }</pre>
 */
public final class MgxClient {

    public static final String DEFAULT_BASE_URL = "https://api.mygrainexchange.com/v1";

    /** OAuth2 + endpoint configuration shared with {@link TokenManager}. */
    public static final class AuthOptions {
        public String clientId;
        public String clientSecret;
        public List<String> scopes;
        public String accessToken;
        public String refreshToken;
        public String tokenUrl;
        public String baseUrl;
    }

    private final InventoryApi inventoryApi;
    private final MarketApi marketApi;
    private final BidsApi bidsApi;
    private final TradesApi tradesApi;
    private final TeamsApi teamsApi;
    private final CashBidsApi cashBidsApi;
    private final WebhooksApi webhooksApi;

    private final Inventory$ inventory = new Inventory$();
    private final Market$ market = new Market$();
    private final Bids$ bids = new Bids$();
    private final Trades$ trades = new Trades$();
    private final Teams$ teams = new Teams$();
    private final CashBids$ cashBids = new CashBids$();
    private final Webhooks$ webhooks = new Webhooks$();

    public MgxClient(AuthOptions options) {
        String baseUrl = options.baseUrl != null ? options.baseUrl : DEFAULT_BASE_URL;

        ApiClient apiClient = new ApiClient();
        apiClient.setBasePath(baseUrl);

        // Use a dedicated, un-intercepted OkHttp client for the token endpoint so
        // token requests never carry (or trigger fetching of) an Authorization header.
        final OkHttpClient tokenHttp = apiClient.getHttpClient().newBuilder().build();
        final TokenManager tokens = new TokenManager(options, baseUrl, tokenHttp);

        // The generated OAuth auth prepends "Bearer " itself when an access token is set,
        // but it only holds a static string. To get correct per-request refresh we own
        // the header via an interceptor and leave the generated auth's token unset.
        Interceptor authInterceptor = new Interceptor() {
            @Override
            public Response intercept(Interceptor.Chain chain) throws IOException {
                Request original = chain.request();
                Request authed = original.newBuilder()
                        .header("Authorization", "Bearer " + tokens.getAccessToken())
                        .build();
                return chain.proceed(authed);
            }
        };
        apiClient.setHttpClient(
                apiClient.getHttpClient().newBuilder().addInterceptor(authInterceptor).build());

        this.inventoryApi = new InventoryApi(apiClient);
        this.marketApi = new MarketApi(apiClient);
        this.bidsApi = new BidsApi(apiClient);
        this.tradesApi = new TradesApi(apiClient);
        this.teamsApi = new TeamsApi(apiClient);
        this.cashBidsApi = new CashBidsApi(apiClient);
        this.webhooksApi = new WebhooksApi(apiClient);
    }

    // ---- builder ----------------------------------------------------------

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private final AuthOptions opts = new AuthOptions();

        public Builder clientId(String v) {
            opts.clientId = v;
            return this;
        }

        public Builder clientSecret(String v) {
            opts.clientSecret = v;
            return this;
        }

        public Builder scopes(String... v) {
            opts.scopes = java.util.Arrays.asList(v);
            return this;
        }

        public Builder scopes(List<String> v) {
            opts.scopes = v;
            return this;
        }

        public Builder accessToken(String v) {
            opts.accessToken = v;
            return this;
        }

        public Builder refreshToken(String v) {
            opts.refreshToken = v;
            return this;
        }

        public Builder tokenUrl(String v) {
            opts.tokenUrl = v;
            return this;
        }

        public Builder baseUrl(String v) {
            opts.baseUrl = v;
            return this;
        }

        public MgxClient build() {
            return new MgxClient(opts);
        }
    }

    // ---- internal helpers -------------------------------------------------

    private static String idempotencyKey(String supplied) {
        return supplied != null ? supplied : UUID.randomUUID().toString();
    }

    @FunctionalInterface
    private interface Call<R> {
        R run() throws ApiException;
    }

    /** Runs a generated call, mapping {@link ApiException} to {@link MgxApiError}. */
    private static <R> R unwrap(Call<R> call) {
        try {
            return call.run();
        } catch (ApiException e) {
            throw MgxApiError.from(e);
        }
    }

    // ---- resource accessors ----------------------------------------------

    /** Browse and filter anonymized inventory. */
    public Inventory$ inventory() {
        return inventory;
    }

    /** Market commodities, prices, and history. */
    public Market$ market() {
        return market;
    }

    /** Read the authenticated team's own bids and act on counter-offers. */
    public Bids$ bids() {
        return bids;
    }

    /** Read the authenticated team's own trades. */
    public Trades$ trades() {
        return trades;
    }

    /** Read the teams the authenticated user belongs to. */
    public Teams$ teams() {
        return teams;
    }

    /** Manage the elevator's own cash bids and the offers received on them. */
    public CashBids$ cashBids() {
        return cashBids;
    }

    /** Subscribe to event notifications and inspect deliveries. */
    public Webhooks$ webhooks() {
        return webhooks;
    }

    // ---- resource wrappers ------------------------------------------------

    /** Inventory resource. */
    public final class Inventory$ {
        public Pagination<Inventory> list() {
            return list(new InventoryListFilters());
        }

        public Pagination<Inventory> list(InventoryListFilters f) {
            final InventoryListFilters filters = f != null ? f : new InventoryListFilters();
            return new Pagination<>(offset -> {
                InventoryPage page = inventoryApi.inventoryList(
                        filters.commodity, filters.grade, filters.minQuantity, filters.maxQuantity,
                        filters.province, filters.near, filters.cropYear, filters.isOrganic,
                        filters.minProtein, filters.sort, filters.limit, offset, null);
                return new Pagination.PageLike<Inventory>() {
                    public List<Inventory> getItems() {
                        return page.getItems();
                    }

                    public String getNext() {
                        return page.getNext();
                    }
                };
            });
        }

        public Inventory get(String id) {
            return unwrap(() -> inventoryApi.inventoryGet(id, null)).getData();
        }

        public InventoryFilters filters() {
            return unwrap(inventoryApi::inventoryFilters);
        }

        public Bid placeBid(String id, PlaceBid body) {
            return placeBid(id, body, null);
        }

        public Bid placeBid(String id, PlaceBid body, String idempotencyKey) {
            return unwrap(() -> bidsApi.bidPlace(id, body, idempotencyKey(idempotencyKey))).getData();
        }
    }

    /** Market resource. */
    public final class Market$ {
        public List<MarketCommodity> commodities() {
            List<MarketCommodity> items = unwrap(() -> marketApi.marketCommodities(null)).getItems();
            return items != null ? items : Collections.emptyList();
        }

        public List<MarketPrice> prices() {
            return prices(null, null);
        }

        public List<MarketPrice> prices(String commodity, LocalDate date) {
            List<MarketPrice> items = unwrap(() -> marketApi.marketPrices(commodity, date, null)).getItems();
            return items != null ? items : Collections.emptyList();
        }

        public PriceHistory history(String commodity) {
            return history(commodity, null, null, null);
        }

        public PriceHistory history(String commodity, LocalDate from, LocalDate to, String interval) {
            return unwrap(() -> marketApi.marketHistory(commodity, from, to, interval, null));
        }
    }

    /** Bids resource. */
    public final class Bids$ {
        public Pagination<Bid> list() {
            return list(null);
        }

        public Pagination<Bid> list(String status) {
            return new Pagination<>(offset -> {
                BidPage page = bidsApi.bidList(status, null, offset, null);
                return new Pagination.PageLike<Bid>() {
                    public List<Bid> getItems() {
                        return page.getItems();
                    }

                    public String getNext() {
                        return page.getNext();
                    }
                };
            });
        }

        public Bid get(String id) {
            return unwrap(() -> bidsApi.bidGet(id, null)).getData();
        }

        public Trade accept(String id) {
            return unwrap(() -> bidsApi.bidAccept(id)).getData();
        }

        public Bid reject(String id) {
            return unwrap(() -> bidsApi.bidReject(id)).getData();
        }

        public Bid counter(String id, PlaceBid body) {
            return unwrap(() -> bidsApi.bidCounter(id, body)).getData();
        }
    }

    /** Trades resource. */
    public final class Trades$ {
        public Pagination<Trade> list() {
            return list(new TradeListFilters());
        }

        public Pagination<Trade> list(TradeListFilters f) {
            final TradeListFilters filters = f != null ? f : new TradeListFilters();
            return new Pagination<>(offset -> {
                TradePage page = tradesApi.tradeList(
                        filters.status, filters.commodity, filters.from, filters.to, filters.limit, offset, null);
                return new Pagination.PageLike<Trade>() {
                    public List<Trade> getItems() {
                        return page.getItems();
                    }

                    public String getNext() {
                        return page.getNext();
                    }
                };
            });
        }

        public Trade get(String id) {
            return unwrap(() -> tradesApi.tradeGet(id, null)).getData();
        }
    }

    /** Teams resource. */
    public final class Teams$ {
        public List<Team> list() {
            List<Team> items = unwrap(teamsApi::teamList).getItems();
            return items != null ? items : Collections.emptyList();
        }

        public Team get(String id) {
            return unwrap(() -> teamsApi.teamGet(id)).getData();
        }
    }

    /** Cash bids resource. */
    public final class CashBids$ {
        public Pagination<CashBid> list() {
            return list(new CashBidListFilters());
        }

        public Pagination<CashBid> list(CashBidListFilters f) {
            final CashBidListFilters filters = f != null ? f : new CashBidListFilters();
            return new Pagination<>(offset -> {
                CashBidPage page = cashBidsApi.cashBidList(filters.isActive, filters.commodity, filters.limit, offset, null);
                return new Pagination.PageLike<CashBid>() {
                    public List<CashBid> getItems() {
                        return page.getItems();
                    }

                    public String getNext() {
                        return page.getNext();
                    }
                };
            });
        }

        public CashBid create(StoreCashBid body) {
            return unwrap(() -> cashBidsApi.cashBidCreate(body)).getData();
        }

        public CashBid update(String id, UpdateCashBid body) {
            return unwrap(() -> cashBidsApi.cashBidUpdate(id, body)).getData();
        }

        public Pagination<CashBidOffer> offers(String cashBidId) {
            return offers(cashBidId, null);
        }

        public Pagination<CashBidOffer> offers(String cashBidId, String status) {
            return new Pagination<>(offset -> {
                CashBidOfferPage page = cashBidsApi.cashBidOffers(cashBidId, status, null, offset);
                return new Pagination.PageLike<CashBidOffer>() {
                    public List<CashBidOffer> getItems() {
                        return page.getItems();
                    }

                    public String getNext() {
                        return page.getNext();
                    }
                };
            });
        }

        public Trade acceptOffer(String offerId) {
            return unwrap(() -> cashBidsApi.cashBidOfferAccept(offerId)).getData();
        }

        public CashBidOffer rejectOffer(String offerId) {
            return unwrap(() -> cashBidsApi.cashBidOfferReject(offerId)).getData();
        }
    }

    /** Webhooks resource. */
    public final class Webhooks$ {
        public List<WebhookResource> list() {
            List<WebhookResource> items = unwrap(webhooksApi::webhookList).getItems();
            return items != null ? items : Collections.emptyList();
        }

        public WebhookCreated create(Webhook body) {
            return unwrap(() -> webhooksApi.webhookCreate(body)).getData();
        }

        public void delete(String id) {
            unwrap(() -> webhooksApi.webhookDelete(id));
        }

        public Pagination<WebhookDelivery> deliveries(String webhookId) {
            return new Pagination<>(offset -> {
                WebhookDeliveryPage page = webhooksApi.webhookDeliveries(webhookId, null, offset);
                return new Pagination.PageLike<WebhookDelivery>() {
                    public List<WebhookDelivery> getItems() {
                        return page.getItems();
                    }

                    public String getNext() {
                        return page.getNext();
                    }
                };
            });
        }

        /**
         * Verify an inbound webhook's signature and return the typed event. Pass the
         * raw request body, the {@code MGX-Signature} header, and the subscription
         * secret. Uses the default tolerance (300 seconds) and the system clock.
         *
         * @throws WebhookVerifier.MgxSignatureError on a bad signature or stale timestamp
         */
        public WebhookEvent verify(String rawBody, String signatureHeader, String secret) {
            return WebhookVerifier.verify(rawBody, signatureHeader, secret);
        }

        /**
         * Verify an inbound webhook's signature and return the typed event, with an
         * explicit tolerance window and optional clock override.
         *
         * @param toleranceSeconds max age of the signed timestamp, in seconds
         * @param now              override for "now" (unix seconds); pass {@code null} for the system clock
         * @throws WebhookVerifier.MgxSignatureError on a bad signature or stale timestamp
         */
        public WebhookEvent verify(
                String rawBody, String signatureHeader, String secret, int toleranceSeconds, Long now) {
            return WebhookVerifier.verify(rawBody, signatureHeader, secret, toleranceSeconds, now);
        }
    }

    // ---- filter holders (overlay-only request params) ---------------------

    /** Fluent filters for {@link Inventory$#list(InventoryListFilters)}. */
    public static final class InventoryListFilters {
        String commodity;
        String grade;
        BigDecimal minQuantity;
        BigDecimal maxQuantity;
        String province;
        String near;
        Integer cropYear;
        Boolean isOrganic;
        BigDecimal minProtein;
        String sort;
        Integer limit;

        public InventoryListFilters commodity(String v) { this.commodity = v; return this; }
        public InventoryListFilters grade(String v) { this.grade = v; return this; }
        public InventoryListFilters minQuantity(BigDecimal v) { this.minQuantity = v; return this; }
        public InventoryListFilters maxQuantity(BigDecimal v) { this.maxQuantity = v; return this; }
        public InventoryListFilters province(String v) { this.province = v; return this; }
        public InventoryListFilters near(String v) { this.near = v; return this; }
        public InventoryListFilters cropYear(Integer v) { this.cropYear = v; return this; }
        public InventoryListFilters isOrganic(Boolean v) { this.isOrganic = v; return this; }
        public InventoryListFilters minProtein(BigDecimal v) { this.minProtein = v; return this; }
        public InventoryListFilters sort(String v) { this.sort = v; return this; }
        public InventoryListFilters limit(Integer v) { this.limit = v; return this; }
    }

    /** Fluent filters for {@link Trades$#list(TradeListFilters)}. */
    public static final class TradeListFilters {
        String status;
        String commodity;
        LocalDate from;
        LocalDate to;
        Integer limit;

        public TradeListFilters status(String v) { this.status = v; return this; }
        public TradeListFilters commodity(String v) { this.commodity = v; return this; }
        public TradeListFilters from(LocalDate v) { this.from = v; return this; }
        public TradeListFilters to(LocalDate v) { this.to = v; return this; }
        public TradeListFilters limit(Integer v) { this.limit = v; return this; }
    }

    /** Fluent filters for {@link CashBids$#list(CashBidListFilters)}. */
    public static final class CashBidListFilters {
        Boolean isActive;
        String commodity;
        Integer limit;

        public CashBidListFilters isActive(Boolean v) { this.isActive = v; return this; }
        public CashBidListFilters commodity(String v) { this.commodity = v; return this; }
        public CashBidListFilters limit(Integer v) { this.limit = v; return this; }
    }
}
